File: ABC/abc002/abc002_4 2.py
def dfs(i: int, members: list[int]) -> int:
    if i == N:
        return len(members)
    
    result = dfs(i+1, members) #議員iを含めない場合

    if all(j in g[i] for j in members):
        members.append(i)
        result = max(result, dfs(i+1, members)) #議員iを含める場合
        members.pop() #元に戻す
    
    return result


N, M = map(int, input().split())

g = [set() for _ in range(N)]
for _ in range(M):
    x, y = map(int, input().split())
    g[x-1].add(y-1)
    g[y-1].add(x-1)

print(dfs(0, []))

File: ABC/abc002/abc002_4 2.cpp
#include <bits/stdc++.h>
using namespace std;

int dfs(int i, int N, const vector<unordered_set<int>>& g, vector<int> members) {
    if (i == N) return members.size();

    int result = dfs(i+1, N, g, members); // 議員iを含めない場合

    bool ok = true; // 議員iが、すでに派閥に含まれる議員の全員と仲が良いかを確認
    for (int j : members) {
        if (g[i].find(j) == g[i].end()) ok = false;
    }
    if (ok) {
        members.push_back(i);
        result = max(result, dfs(i+1, N, g, members)); // 議員iを含める場合
    }

    return result;
}

int main() {
    int N, M, x, y;
    cin >> N >> M;
    
    vector<unordered_set<int>> g(N, unordered_set<int>{});
    for (int i = 0; i < M; i++) {
        cin >> x >> y;
        g[x-1].insert(y-1);
        g[y-1].insert(x-1);
    }

    cout << dfs(0, N, g, {}) << endl;
}